src/transformers/NestedLoopTransformation.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2024, Konstantin Britikov <britikovki@gmail.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "NestedLoopTransformation.h"
+
+#include "QuantifierElimination.h"
+#include "TransformationUtils.h"
+#include "graph/ChcGraph.h"
+#include "utils/SmtSolver.h"
+
+std::unique_ptr<NestedLoopTransformation::WitnessBackTranslator>
+NestedLoopTransformation::transform(ChcDirectedGraph & graph) {
+    auto vertices = graph.getVertices();
+    auto originalGraph = graph;
+    std::vector<EId> loop = detectLoop(graph);
+    std::vector<WitnessInfo> loops;
+    while (loop.size() > 1) {
+        auto wtnsInfo = graph.contractConnectedVertexes(loop);
+        loops.push_back(wtnsInfo);
+        loop = detectLoop(graph);
+    }
+    return std::make_unique<NestedLoopTransformation::WitnessBackTranslator>(std::move(originalGraph), graph, loops);
+}
+
+VerificationResult NestedLoopTransformation::WitnessBackTranslator::translate(VerificationResult & result) {
+    if (result.getAnswer() == VerificationAnswer::UNSAFE) {
+        auto witness = translateErrorPath(result.getInvalidityWitness());
+        if (std::holds_alternative<NoWitness>(witness)) {
+            return {result.getAnswer(), std::get<NoWitness>(std::move(witness))};
+        }
+        return {VerificationAnswer::UNSAFE, std::get<InvalidityWitness>(std::move(witness))};
+    }
+    if (result.getAnswer() == VerificationAnswer::SAFE) {
+        auto witness = translateInvariant(result.getValidityWitness());
+        if (std::holds_alternative<NoWitness>(witness)) {
+            return {result.getAnswer(), std::get<NoWitness>(std::move(witness))};
+        }
+        return {VerificationAnswer::SAFE, std::get<ValidityWitness>(std::move(witness))};
+    }
+    return std::move(result);
+}
+
+NestedLoopTransformation::WitnessBackTranslator::ErrorOr<InvalidityWitness>
+NestedLoopTransformation::WitnessBackTranslator::translateErrorPath(InvalidityWitness wtns) {
+    // We need to get the CEX path, which will define the locations in the graph
+    Logic & logic = graph.getLogic();
+    TimeMachine tm(logic);
+    auto errorPath = wtns.getDerivation();
+    std::vector<SymRef> pathVertices;
+    pathVertices.push_back(initialGraph.getEntry());
+    auto newVertices = graph.getVertices();
+    auto allVertices = initialGraph.getVertices();
+
+    // Compute model for the error path
+    auto model = [&]() {
+        SMTSolver solverWrapper(logic, SMTSolver::WitnessProduction::ONLY_MODEL);
+        auto & solver = solverWrapper.getCoreSolver();
```

Comment:
There is no longer a need to access the wrapped solver.
You can achieve what you need with the interface of `SMTSolver` now.